lib/algolia/middleware/response.ex

```diff
@@ -0,0 +1,17 @@
+defmodule Algolia.Middleware.Response do
```

Comment:
It kind of seems like this library invented the [strange response types](https://github.com/sikanhe/algolia-elixir?tab=readme-ov-file#return-values) with different tuple sizes and different types for the second element.

This would be surprising for normal users of Tesla and make reported events to Tesla.Middleware.Telemetry inconsistent if an end user was using Tesla.Middleware.Telemetry for other APIs (including Slab).

I think this update could be the opportunity to make very backwards incompatible changes (which is really just adhering to Tesla's native `request` return types) and maybe even make a clean break from the original library.

Reply:
This middleware should be placed in a position in the chain where the telemetry events are unaffected by it, as it will modify the result after `Telemetry` has already processed it.

But even so, it sounds like you think we should just stop unwrapping the `Tesla.Env` entirely, so `:error` tuples are returned only for connection-level errors, and HTTP errors are represented as `:ok` responses, where checking the status code is needed.

I think that would be fine to do. One complication: this library injects the index name onto responses that don't already have it, largely to make it possible to do `wait_task` via pipe. I think it's weird to alter the response in that way, so instead I think we should include it in the `opts` of the request, and then `wait_task` can pull it off the `env`, now that it has that full context.

Reply:
How indexes are handled is another poor API decision from the original library in my view. Most APIs require it to be passed, even though the most common use case would be one index. Even if a user needed multiple, they could easily just create multiple clients. But most APIs take a client and an index, elevating the use case of same client for multiple different indexes as the primary use. And then there are APIs like `wait` that don't take an index this injecting behavior to make it happen. There are also miscellaneous decisions like supporting [environment variable configurations](https://github.com/sikanhe/algolia-elixir?tab=readme-ov-file#using-environment-variables) when the Config docs literally warn [not to do this](https://share.slab.com/RTpH64CG).

Not sure when is the best time (as in before or after this PR) but this is why I think entirely changing the API is on the table.

Reply:
I'm also in favor of moving the index into the client, so that you need to use distinct clients if you use multiple indices. I would rather do it in another PR though, so do you mind if we keep the `Response` middleware for this one, just as a temporary shim to keep the same responses as before? And then I'll remove it in a subsequent PR that will give normal Tesla responses and make the index choice be part of the client.